rfcs/000-search-revamp.md

```diff
@@ -0,0 +1,124 @@
+- Feature Name: Search Revamp
+- Start Date: 2021-09-20
+- RFC PR: todo
+
+# Search Revamp
+
+## Summary
+
+Revamp search with a cleaner backend implementation, more robust queries, better customizability, and
+making use of modern v7 ES features.  
+
+## Motivation
+
+At Lyft, search enhancement are the #1 most frequent Amundsen feedback. This is no surprise, as the current 
+technical state of search has not changed much from MVP.
+
+We see search as serving two distinct (but related) use cases:
+  - known-data searches, which have a starting point (e.g. tables with a `user_id` column in the `production` schema) 
+    and are more interested in listing the available data matching a known condition
+  - discovery searches, which are looking to find previously-unknown data that would be relevant for solving
+    one's problem
+
+Although we believe Amundsen has been particularly lacking in the discovery search category, our proposal is seeking
+to improve the experience of both search types. 
+
+
+## UI/UX Details
+
+Although the bulk of proposed changes will be to search result quality and relevance, we're also proposing
+a handful of UI/UX changes to filters. We're currently working on designs and will share those as soon as they're
+available.
+
+#### Multiple Values Per Filter.
+Allow multiple values (likely comma-separated) for filters, which has been very highly requested. 
+
+There is an obvious question of whether multiple values would be AND'ed or OR'ed together. Our goal is to provide
+additional expressiveness without going down the path of supporting full logical expressions. To better understand the problem,
+keep in mind that there's two kind of fields: single-valued fields (table name, schema name) and multi-valued fields (columns, tags, badges).
+Our proposed tradeoff between expressiveness and simplicity is to combine filter values for single-valued 
+fields with OR (e.g. "staging,production" filter applied to the "schema" field)
+whereas for multi-valued fields we'd give users an options to choose AND or OR 
+(a user would choose AND/OR to apply the filters "product_id,user_id" to the "column" field)
+
+#### Negation filtering
+Another highly requested feature is specifying negative filters - e.g. "schema is not temp, tmp, or temporary".
```

Comment:
Apart from negation, positive boosting is also valid use-case. In ING we hard-coded it on query level (which could be something configurable in search proxy) to promote tables coming from official data sources. So if the table has `ing_official_data` tag it's relevance is boosted in the search results.

This could be something simple like configuration option - list of maps with field, value, boost_factor properties.

Reply:
Yes! We do something very very similar at Lyft. I also left this out of the RFC (since it's more of a user-configurable feature). Our plan, once queries are composed of extendable native python objects and not templated JSON, is to use [Rescoring](https://www.elastic.co/guide/en/elasticsearch/reference/7.2/search-request-rescore.html) for this. I.e. you'd provide a `rescore` query to match docs that have the `ing_official_data` tag and then provide coefficients of how to combine the score of the `rescore` query vs the original query. 

FYI rescoring isn't natively supported by the python lib but it can be injected as an `extra`, see [issue](https://github.com/elastic/elasticsearch-dsl-py/issues/311).

Reply:
hm the idea with rescoring is fine idea but limiting in a sense that it doesn't span `all` of the results but only top N (10 by default) from each shard (afair there is only 1 shard by default in our mappings and our use case will very rarely require more than that). so `negating` or `promoting` would only touch top 10 of results from the original es scoring.

Reply:
I had figured this was not much of a limitation as one could adjust N to a point at which one can say "hey, if the result wasn't in the top 500 to begin with, maybe it shouldn't get boosted at all". But we're not married to this implementation - we're looking to let users tweak scoring according to their custom logic in a clean and extendable way (that plays well with the original document score instead of just overriding it) and we're open to suggestions.